event/parsing/eventValidator.go

```diff
@@ -0,0 +1,146 @@
+package parsing
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/xmidt-org/glaukos/event/client"
+	"github.com/xmidt-org/glaukos/event/queue"
+)
+
+// EventValidation is an interface for checking if wrp messages and events pass certain validations
+type EventValidation interface {
+	IsEventValid(client.Event) (bool, error)
+	IsWRPValid(queue.WrpWithTime) (bool, error)
+	GetEventCompareTime(client.Event) (time.Time, error)
+	GetWRPCompareTime(queue.WrpWithTime) (time.Time, error)
+	ValidateType(string) bool
+}
+
+// EventRule is the config struct for event validation
+type EventRule struct {
+	Regex          string
+	CalculateUsing string
+	ValidFrom      time.Duration
+}
+
+// eventValidator implements a metricparser.EventValidator interface.
+// Keeps a set of information/rules to see if an event or wrp fits these rules.
+type eventValidator struct {
+	regex          *regexp.Regexp
+	calculateUsing TimeLocation
+	timeValidation TimeValidation
+}
+
+var (
+	ErrInvalidEventType = errors.New("event type doesn't match")
+	ErrInvalidRegex     = errors.New("invalid regex")
+	ErrInvalidBootTime  = errors.New("invalid boot-time")
+	ErrInvalidBirthdate = errors.New("invalid birthdate")
+	ErrTimeParse        = errors.New("parsing error")
+)
+
+// NewEventValidation creates a new EventValidator from an EventRule
+func NewEventValidation(rule EventRule, validTo time.Duration, currentTime func() time.Time) (EventValidation, error) {
+	regex, err := regexp.Compile(rule.Regex)
+	if err != nil {
+		return eventValidator{}, fmt.Errorf("%w: %v", ErrInvalidRegex, err)
+	}
+
+	tv := TimeValidator{Current: currentTime, ValidFrom: rule.ValidFrom, ValidTo: validTo}
+
+	timeLocation := ParseTimeLocation(rule.CalculateUsing)
+
+	return eventValidator{
+		regex:          regex,
+		calculateUsing: timeLocation,
+		timeValidation: tv,
+	}, nil
+}
+
+// IsEventValid checks if an event is valid based on the rules kept by the EventValidator.
+func (e eventValidator) IsEventValid(event client.Event) (bool, error) {
+	bootTime, err := GetEventBootTime(event)
+	if err != nil || bootTime <= 0 {
+		return false, ErrInvalidBootTime
+	}
+
+	// see if event found matches expected event type
+	if !e.ValidateType(event.Dest) {
+		return false, fmt.Errorf("%w. Desired type: %s", ErrInvalidEventType, e.regex.String())
+	}
+
+	compareTime, err := e.GetEventCompareTime(event)
+	if err != nil {
+		return false, fmt.Errorf("%w: %v", ErrTimeParse, err)
+	}
+
+	if valid, err := e.timeValidation.IsTimeValid(compareTime); !valid {
+		if e.calculateUsing == Birthdate {
+			return false, fmt.Errorf("%w: %v", ErrInvalidBirthdate, err)
+		}
+		return false, fmt.Errorf("%w: %v", ErrInvalidBootTime, err)
+	}
+
+	return true, nil
+}
+
+// IsWRPValid checks if a wrp is valid baased on the rules kept by the EventValidator.
+func (e eventValidator) IsWRPValid(wrpWithTime queue.WrpWithTime) (bool, error) {
```

Comment:
I'm starting to think `eventValidator` doesn't really describe what this struct is doing? A few options:
- split up the validator into an `eventValidator` and `wrpValidator`
- somehow add `.Validate()` functions to the actual structs (haven't figured this one out yet - especially since we have configurable limits and shared functions betweeen them)
- drop the `event` from the name (really seems like the easiest option here)

The only good reason I can think of to split this into two different structs (that will have a lot of the same dependencies for time reasons) is if a consumer might only need to validate events or only need to validate wrps.

Reply:
Also - you could shorten the function name for brevity's sake to `ValidWRP` (and similarly `ValidEvent`).  I'm more ambivalent about this.

Reply:
Yeah I think I named it `eventValidator` because there's also the `TimeValidator` struct so I wanted to differentiate the two. But I think I'll change `eventValidator` to `rule` instead.